### DISCUSSION TITLE
Saturation with REFPROP Backend

### DISCUSSION THREAD

**Discussion by eliesaikali** [CoolProp/CoolProp#2234] [Q&A]
Hello !

Is it possible to calculate the saturation + its derivative with the REFPROP backend ?

I have the following code : 

CoolProp::AbstractState * fluide_cool = CoolProp::AbstractState::factory("REFPROP", "Water");
 fluide_cool->update(CoolProp::PQ_INPUTS,  P, 0);  // SI units
 rho_sat = fluide_cool->saturated_liquid_keyed_output(CoolProp::iDmass);
 T_sat = fluide_cool->saturated_liquid_keyed_output(CoolProp::iT);

it works for rho but not T

In addition, 
DrhoDP = fluide_cool->first_saturation_deriv(CoolProp::iDmass, CoolProp::iP);

does not work even for density ...

Regards

**Comment by ibell** [comment]
Getting the temperature via keyed_output on one of the saturation phases
(or any other valid output) should definitely work. What error do you get?
In your case, it would be sufficient to do

T_sat = fluide_cool->T();

because the temperature is the same in both phases by definition. For the
REFPROP backend, the saturation derivatives are not implemented. In
principle it is easy to implement, but nobody has stepped up to do it.

**Reply by eliesaikali** [reply]
Hello !
Getting the temperature via the keyed output from the saturation is not possible, non ...
The error is clear from method REFPROPMixtureBackend::calc_saturated_liquid_keyed_output, line 2065 : 
            **throw ValueError("Invalid parameter. Only mass and molar density are available with RefProp");**

Only density output is possible ...


so yes, i did finally PQ_INPUTS and got T_sat, h_sat from T(), and Hmass(), ... etc ... 
same for cp_sat ..

Thats not an issue !

Otherwise, I finally implemented the derivatives via a finite difference method for all saturation properties :) it works fine

same for derivatives of mu, lambda (w.r.t. p and T) which was not also possible via the method first_partial_deriv(...)

Thanks a lot for your help and for your reply !
